From fd30a7ebce00fe3088fd14d5490079f97137a866 Mon Sep 17 00:00:00 2001
From: Lenka Doudova <mirielka@users.noreply.github.com>
Date: Tue, 16 Sep 2025 18:07:10 +0200
Subject: [PATCH] Issue 6753 - Port ticket 49386 test (#6987)

Description:
Port ticket 49386 test into
dirsrvtests/tests/suites/memberof_plugin/regression_test.py

Relates: #6753
Author: Lenka Doudova
Assisted by: Cursor
Reviewer: Thierry Bordaz
---
 .../suites/memberof_plugin/regression_test.py |  52 +++++-
 dirsrvtests/tests/tickets/ticket49386_test.py | 159 ------------------
 2 files changed, 51 insertions(+), 160 deletions(-)
 delete mode 100644 dirsrvtests/tests/tickets/ticket49386_test.py

diff --git a/dirsrvtests/tests/suites/memberof_plugin/regression_test.py b/dirsrvtests/tests/suites/memberof_plugin/regression_test.py
index 0a2f278b4..9b073e694 100644
--- a/dirsrvtests/tests/suites/memberof_plugin/regression_test.py
+++ b/dirsrvtests/tests/suites/memberof_plugin/regression_test.py
@@ -28,7 +28,7 @@ from lib389.idm.domain import Domain
 from lib389.dirsrv_log import DirsrvErrorLog
 from lib389.dseldif import DSEldif
 from contextlib import suppress
-
+from . import check_membership
 
 # Skip on older versions
 pytestmark = [pytest.mark.tier1,
@@ -1443,6 +1443,56 @@ def test_shutdown_on_deferred_memberof(topology_st, request):
     request.addfinalizer(fin)
 
 
+def test_memberof_modrdn_to_itself(topology_st, user1, group1):
+    """Test that memberOf plugin correctly handles modrdn operations on groups
+
+    :id: a4c7f8e9-2f1a-11ef-8a3c-482ae39447e5
+    :setup: Standalone Instance with user1 and group1 fixtures
+    :steps:
+        1. Enable the memberOf plugin and restart the server
+        2. Verify user is not initially a member of the group
+        3. Add user to the group
+        4. Verify user is now a member of the group
+        5. Enable plugin logging to capture modrdn operations
+        6. Perform modrdn operation on group (rename to itself)
+        7. Verify user is still a member of the group after rename
+        8. Verify memberOf plugin logs skip message for identical src/dst rename
+    :expectedresults:
+        1. MemberOf plugin should be enabled successfully
+        2. User should not have memberOf attribute initially
+        3. User should be added to group successfully
+        4. User should have memberOf attribute pointing to group
+        5. Plugin logging should be enabled successfully
+        6. Modrdn operation should complete without errors
+        7. User should retain memberOf attribute after rename operation
+        8. Plugin should log that modrdn was skipped due to identical src/dst
+    """
+
+    # Enable the MemberOf plugin
+    memberof = MemberOfPlugin(topology_st.standalone)
+    memberof.enable()
+    topology_st.standalone.restart()
+
+    # Verify that the user is not a member of the group
+    check_membership(user1, group1.dn, False)
+    # Add the user to the group
+    group1.add_member(user1.dn)
+    # Verify that the user is a member of the group
+    check_membership(user1, group1.dn, True)
+
+    # Enable the plugin log to capture memberof modrdn callback notification
+    topology_st.standalone.config.loglevel(vals=[LOG_PLUGIN, LOG_DEFAULT], service='error')
+
+    # Rename the group on itself
+    group1.rename('cn=group1')
+
+    # Verify that the user is still a member of the group
+    check_membership(user1, group1.dn, True)
+
+    # Verify that the memberof modrdn callback notification is logged
+    assert topology_st.standalone.ds_error_log.match('.*Skip modrdn operation because src/dst identical.*')
+
+
 if __name__ == '__main__':
     # Run isolated
     # -s for DEBUG mode
diff --git a/dirsrvtests/tests/tickets/ticket49386_test.py b/dirsrvtests/tests/tickets/ticket49386_test.py
deleted file mode 100644
index c6a59ea6a..000000000
--- a/dirsrvtests/tests/tickets/ticket49386_test.py
+++ /dev/null
@@ -1,159 +0,0 @@
-# --- BEGIN COPYRIGHT BLOCK ---
-# Copyright (C) 2022 Red Hat, Inc.
-# All rights reserved.
-#
-# License: GPL (version 3 or any later version).
-# See LICENSE for details.
-# --- END COPYRIGHT BLOCK ---
-#
-import logging
-import pytest
-import os
-import ldap
-import time
-from lib389.utils import *
-from lib389.topologies import topology_st as topo
-from lib389._constants import *
-from lib389.config import Config
-from lib389 import Entry
-
-pytestmark = [pytest.mark.tier2,
-              pytest.mark.skipif(ds_is_older('1.3.7'), reason="Not implemented")]
-
-USER_CN='user_'
-GROUP_CN='group_'
-
-DEBUGGING = os.getenv("DEBUGGING", default=False)
-if DEBUGGING:
-    logging.getLogger(__name__).setLevel(logging.DEBUG)
-else:
-    logging.getLogger(__name__).setLevel(logging.INFO)
-log = logging.getLogger(__name__)
-
-def add_user(server, no, desc='dummy', sleep=True):
-    cn = '%s%d' % (USER_CN, no)
-    dn = 'cn=%s,ou=people,%s' % (cn, SUFFIX)
-    log.fatal('Adding user (%s): ' % dn)
-    server.add_s(Entry((dn, {'objectclass': ['top', 'person', 'inetuser'],
-                             'sn': ['_%s' % cn],
-                             'description': [desc]})))
-    if sleep:
-        time.sleep(2)
-
-def add_group(server, nr, sleep=True):
-    cn = '%s%d' % (GROUP_CN, nr)
-    dn = 'cn=%s,ou=groups,%s' % (cn, SUFFIX)
-    server.add_s(Entry((dn, {'objectclass': ['top', 'groupofnames'],
-                             'description': 'group %d' % nr})))
-    if sleep:
-        time.sleep(2)
-        
-def update_member(server, member_dn, group_dn, op, sleep=True):
-    mod = [(op, 'member', ensure_bytes(member_dn))]
-    server.modify_s(group_dn, mod)
-    if sleep:
-        time.sleep(2)
-        
-def config_memberof(server):
-
-    server.plugins.enable(name=PLUGIN_MEMBER_OF)
-    MEMBEROF_PLUGIN_DN = ('cn=' + PLUGIN_MEMBER_OF + ',cn=plugins,cn=config')
-    server.modify_s(MEMBEROF_PLUGIN_DN, [(ldap.MOD_REPLACE,
-                                          'memberOfAllBackends',
-                                          b'on'),
-                                          (ldap.MOD_REPLACE, 'memberOfAutoAddOC', b'nsMemberOf')])
-
-
-def _find_memberof(server, member_dn, group_dn, find_result=True):
-    ent = server.getEntry(member_dn, ldap.SCOPE_BASE, "(objectclass=*)", ['memberof'])
-    found = False
-    if ent.hasAttr('memberof'):
-
-        for val in ent.getValues('memberof'):
-            server.log.info("!!!!!!! %s: memberof->%s" % (member_dn, val))
-            server.log.info("!!!!!!! %s" % (val))
-            server.log.info("!!!!!!! %s" % (group_dn))
-            if val.lower() == ensure_bytes(group_dn.lower()):
-                found = True
-                break
-
-    if find_result:
-        assert (found)
-    else:
-        assert (not found)
-        
-def test_ticket49386(topo):
-    """Specify a test case purpose or name here
-
-    :id: ceb1e2b7-42cb-49f9-8ddd-bc752aa4a589
-    :setup: Fill in set up configuration here
-    :steps:
-        1. Configure memberof
-        2. Add users (user_1)
-        3. Add groups (group_1)
-        4. Make user_1 member of group_1
-        5. Check that user_1 has the memberof attribute to group_1
-        6. Enable plugin log to capture memberof modrdn callback notification
-        7. Rename group_1 in itself
-        8. Check that the operation was skipped by memberof
-        
-    :expectedresults:
-        1. memberof modrdn callbackk to log notfication that the update is skipped
-    """
-
-    S1 = topo.standalone
-    
-    # Step 1
-    config_memberof(S1)
-    S1.restart()
-    
-    # Step 2
-    for i in range(10):
-        add_user(S1, i, desc='add on S1')
-    
-    # Step 3
-    for i in range(3):
-        add_group(S1, i)
-    
-    # Step 4
-    member_dn = 'cn=%s%d,ou=people,%s' % (USER_CN,  1, SUFFIX)
-    group_parent_dn = 'ou=groups,%s' % (SUFFIX)
-    group_rdn = 'cn=%s%d' % (GROUP_CN, 1)
-    group_dn  = '%s,%s' % (group_rdn, group_parent_dn)
-    update_member(S1, member_dn, group_dn, ldap.MOD_ADD, sleep=False)
-    
-    # Step 5
-    _find_memberof(S1, member_dn, group_dn, find_result=True)
-    
-    # Step 6
-    S1.config.loglevel(vals=[LOG_PLUGIN, LOG_DEFAULT], service='error')
-    
-    # Step 7
-    S1.rename_s(group_dn, group_rdn, newsuperior=group_parent_dn, delold=0)
-    
-    # Step 8
-    time.sleep(2) # should not be useful..
-    found = False
-    for i in S1.ds_error_log.match('.*Skip modrdn operation because src/dst identical.*'):
-        log.info('memberof log found: %s' % i)
-        found = True
-    assert(found)
-    
-    # If you need any test suite initialization,
-    # please, write additional fixture for that (including finalizer).
-    # Topology for suites are predefined in lib389/topologies.py.
-
-    # If you need host, port or any other data about instance,
-    # Please, use the instance object attributes for that (for example, topo.ms["supplier1"].serverid)
-
-    if DEBUGGING:
-        # Add debugging steps(if any)...
-        pass
-
-
-if __name__ == '__main__':
-    # Run isolated
-    # -s for DEBUG mode
-    CURRENT_FILE = os.path.realpath(__file__)
-    pytest.main("-s %s" % CURRENT_FILE)
-
-- 
2.52.0

